### PR TITLE
Upgrade to Electron 15

### PR DESCRIPTION
**Pull request by krassowski**
Looks like this one will be easy. Electron 15 blog post: https://www.electronjs.org/blog/electron-15-0
We are getting straight to 15.3.4.

Full changelog for 15.0.0: https://www.electronjs.org/releases/stable?page=3#15.0.0

Stack:
- Chromium 93 → 94
- Node v14.17.0 → v16.5.0
- V8: v9.3 → [v9.4](https://v8.dev/blog/v8-release-94)

### CODE REVIEW

**Review by mbektas — APPROVED** [review]
Great. Thanks!

**Comment by mbektas** [comment]
I am getting kernel related errors once I install and test on my Mac. Released version 3.2.5-2 works fine on the same machine.

electron 15 problem

**Comment by krassowski** [comment]
I can reproduce `Error during WebSocket handshake: Unexpected response code 403`. Nothing I tried so far helped:
- `nativeWindowOpen: false` in `BrowserWindow` constructor
- `--ServerApp.allow_remote_access=True`
- `--log_level=DEBUG` to check if the message gets received by the server and if there is more logs (it does not seem to be)
- downgrading Electron to 15.0.0
- setting custom CORS

I will appreciate any suggestions on how to tackle this one.

**Comment by mbektas** [comment]
@krassowski can you rebase your branch so we can get test binaries with changes in https://github.com/jupyterlab/jupyterlab-desktop/pull/387?

**Comment by krassowski** [comment]
Done

**Comment by mbektas** [comment]
Thanks @krassowski . I tested on Mac and noticing two issues:
- PDF documents are not opening anymore
- After switching to Classic notebooks and clicking on a notebook opens a blank window
could you look into these please?

**Comment by mbektas** [comment]
> Thanks @krassowski . I tested on Mac and noticing two issues:
> 
> * PDF documents are not opening anymore
> * After switching to Classic notebooks and clicking on a notebook opens a blank window
>   could you look into these please?

just checked the same with https://github.com/jupyterlab/jupyterlab-desktop/pull/387 and both work fine with those changes. so that makes me think these are failing maybe because of additional security measures in electron 15?

**Comment by mbektas** [comment]
@krassowski `atob` change was necessary to get PDFs to open properly. PDF content is returned in base64 format from contents API but it contains whitespaces. This base64 content is converted to binary blob to render. atob function is used in the process. External browser's implementation of atob permits whitespaces. Since Desktop app creates BrowserWindow [without context isolation](https://github.com/jupyterlab/jupyterlab-desktop/blob/master/src/main/sessions.ts#L406), atob from node.js is exposed to browser window. atob of node.js does not permit whitespaces and causes PDF blob generation to fail. with my changes I am introducing a modified version of atob of node.js which allows whitespaces. Once we turn on contextIsolation, this patch wouldn't be needed anymore.

File: src/main/main.ts [hunk 1]
```diff
@@ -190,10 +190,13 @@ app.on('ready', () => {
 });
 
 app.on("web-contents-created", (_event: any, webContents: WebContents) => {
-    // Prevent navigation
+    // Prevent navigation to local links
```
Comment (mbektas) [inline src/main/main.ts:193]: @krassowski do you have any concerns on these changes? any concerns navigating to external websites?

**Comment by krassowski** [comment]
> atob function is used in the process. External browser's implementation of atob permits whitespaces. Since Desktop app creates BrowserWindow [without context isolation](https://github.com/jupyterlab/jupyterlab-desktop/blob/master/src/main/sessions.ts#L406), atob from node.js is exposed to browser window.

Was it a change in Electron 15 which lead to this?

File: src/main/main.ts (hunk 1 above)
Comment (krassowski) [reply src/main/main.ts:193]: Yes, I am concerned. Wouldn't this lead to a context leak? Why is this change needed, again?

**Comment by mbektas** [comment]
> 

yes, atob issue turns out to be caused by Electron 15 upgrade.

File: src/main/main.ts (hunk 1 above)
Comment (mbektas) [reply src/main/main.ts:193]: during my testing, external domain urls were opened in popup windows and that is safe, so I thought we could enable it back. but if you think there are cases that require blocking external urls then I can update it to block. The original motivation for this change was the issue I mentioned in earlier comments: `After switching to Classic notebooks and clicking on a notebook opens a blank window`

**Comment by krassowski** [comment]
How about https://github.com/krassowski/jupyterlab-desktop/pull/15 ?

**Review by mbektas — APPROVED** [review]

**Pull request merged**